### ISSUE TITLE
Feedback: Off-Mesh Links

### ISSUE THREAD

**Issue by DarrelCusey** [Unity-Technologies/NavMeshComponents#11]
These improvements look great, but there are three issues with Unity's Navigation system that have been open and pending for a LONG time that I hope will be addressed before these NEW features are added:

1) Can we have an option to make all automatically-generated off-mesh links always bi-directional? Currently, when an off-mesh link is created, it will ONLY be bi-directional if the elevation change between the two segments is less than some very small amount (creating a "jump" link).  Or, even better, expose a float field that can be used in the calculation above to multiply the range of the allowed elevation change to make a bi-directional "jump" links instead of just a "drop-down" link.

2) Can the automatically-generated off-mesh link data be made available in the API?  When a NavMesh is created, a "NavMesh" object should be created in the Scene Hierarchy, which should contain all NavMesh segments and all Off-Mesh Links.  If we could get access to these components at run-time, it would allow us to do things like fixing/aligning the vertices of NavMesh segments, or setting the Off-Mesh links to be bi-directional.

3)  Can we have an option to set the SPEED at which off-mesh links are traversed be dependent on the LENGTH of the off-mesh link?  Currently, all off-mesh links are traversed in about 1 second, no matter how long those links are.  It looks like the off-mesh links are always assumed to be about 3 meter long, and so when off-mesh links are much longer, the NavMeshAgent traverses the links so fast that it looks like the NaveMeshAgent is teleporting over the link.  If this option was set to true, then a "speed modifier" would be set in the OffMeshLink data that would be used to modify the speed of the NavMeshAgent while it is traversing the link.  The NavMeshAgent speed would be restored to its original value after off-mesh link traversal is completed.

**Comment by jakobhu** [comment]
We plan to replace the auto generation of offmeshlinks with something better and more tweakable. This is not currently in the works though - but would eventually cover 1, 2.

For 3) - The actual behavior is using twice the regular speed of the agent - considering the horizontal distance only - see https://docs.unity3d.com/ScriptReference/AI.NavMeshAgent.CompleteOffMeshLink.html for how to customize link movement.

**Comment by HilariousCow** [comment]
I'm currently working on doing my own auto generation of navmesh links. The new NavMeshLink is nice, and I like the way it has a width. But the width is actually only really useful between parallel region edges.

It would be quite nice to be able to define all the vertices of a NavMeshLink Quad... build a little bridge, as it were. But at that point I'm kind of just augmenting the mesh with more mesh. It's just that I don't have the ability to upload the mesh I've extracted back into the NavMesh, plus some area data per triangle or whatever. 

I guess I could place AreaModifiers over the extra NavMesh segments I'd be uploading? But then, those only have certain shapes, so I can't do it on a per vert/face basis.

Sorry, just thinking out loud. 

tl:dr: NavMeshLinkPoly would be a really nice thing.

**Comment by DKaravolos** [comment]
I would love to have automatic off-mesh link generation (mostly for agents dropping down from procedurally generated ledges) without having to use the UnityEditor. Are there any updates on this? Any idea when this new system will be implemented? 
Looking forward to it!

**Comment by HilariousCow** [comment]
We've started moving back to A* Pathfinding Project because of issues we encountered wrt NavMeshLink/OffMeshLink. We didn't think that these would be addressed within our time frame.

The custom code we wrote to automate the generation of links works fine, but currently, it's not possible to determine if a path returned by the Unity NavMesh system incorporated a NavMeshLink/OffMeshLink into the path. It is simply a set of anonymous corners. The only way to detect a NavMeshLink/OffMeshLink as part of a path is to step a NavMeshAgent along a path, and check if a bool comes on, stating that you're on one of these. I.e. You can't tell that you've been given a path with a "jump this gap" section in until you've already jumping the gap.

As a result, you can't tweak your path steering code to accommodate pre-emptive actions (unless you create a dummy agent to walk the path ahead of you, which, seems... uh...).

...

Thankfully the work we did with NavMeshLink generation, (extracting non-manifold edge region loops from the nav mesh, and then comparing edges against one another) were still applicable to A* Pathfinding Project, as they both use Recaster to generate Nav Meshes from arbitrary geometry (afaik), so the results are very similar.

We did, however, have to do a lot of "tidying" of the meshes because of the way the scene is split into tiles - you get back a mesh with duplicate vertices and holes along tile edges. Again, this was true for both Unity and A* Pathfinding Project.

The other thing I would note is that the reach-abilities are extremely tied to the particular implementation of your character movers, which is something that would fall outside of Unity's purview, as the standard character controller is a bit basic. We use KinematicCharacterController from the app store (best general purpose controller I've seen so far), plus some tweaks for more forgiving step-up heights, and we use the "impact normal" (i.e. face normal) rather than "edge normal" of a ground probe to determine if the character is standing on the ground. I mention these because they have to be accounted for when we probe for "ledge to ledge" reachability and "drop reachablility" (i.e. if you have an overhang there's no "floor ledge" to construct a Link to).  We also have to be very careful when making Sweep checks from a region, because the edges do not necessarily run along the geometry that constructed them - they're approximate. We do extra probes along the stretch of the edge to find the "true ground", then find a point projected along the surface normal that hangs over the "true edge" of the ground, and do a swept trace from there using the character dimensions.

In short, it's a lot of work, and also, it's hard to believe a decent general solution could be made without making code far more transparent.

**Comment by DKaravolos** [comment]
Thank you for the extensive reply! The current system is pretty cool as it is, thanks for that.  I found a workaround for my own project, so I'm okay. Good luck with the A* Pathfinding Project!

**Comment by wilg** [comment]
@jakobhu Is it possible to use the existing "Generate Off-Mesh Links" behavior with the component workflow? I can't find a way. I'm a bit hamstrung in my project because I need to use the component workflow but its not really feasible to manually build the off mesh links.

**Comment by turadr** [comment]
@wilg Unfortunately it is still not possible to auto-generate the links with the component workflow. 

(Also mentioned in the last question of the [FAQ](https://github.com/Unity-Technologies/NavMeshComponents#faq) )

**Comment by eDmitriy** [comment]
Hello! I'm working on Navmesh Link Generator https://forum.unity.com/threads/navmesh-links-generator-for-navmeshcomponents.515143/ 
Can someone help me with testing?

**Comment by thecrazy** [comment]
> We plan to replace the auto generation of offmeshlinks with something better and more tweakable. This is not currently in the works though - but would eventually cover 1, 2.
> 
> For 3) - The actual behavior is using twice the regular speed of the agent - considering the horizontal distance only - see https://docs.unity3d.com/ScriptReference/AI.NavMeshAgent.CompleteOffMeshLink.html for how to customize link movement.

Hello, paying customer here. Without point 1 I will have to change my navemesh solution... that takes time. MUCH MUCH more than adding a bool to flip "NavMeshLinkData.bidirectional" to true before baketime. Unless there is something else involved this is a 15 min project. PLEASE put someone on it.

**Comment by JuhanaKivela** [comment]
> 1. Can we have an option to make all automatically-generated off-mesh links always bi-directional? Currently, when an off-mesh link is created, it will ONLY be bi-directional if the elevation change between the two segments is less than some very small amount (creating a "jump" link).  Or, even better, expose a float field that can be used in the calculation above to multiply the range of the allowed elevation change to make a bi-directional "jump" links instead of just a "drop-down" link.

Also hoping for an update for the number 1. Manually putting jump links will take quite a lot of time and the bool "make it bi-directional" would solve this totally.

**Comment by thecrazy** [comment]
@jakobhu Any update on this? If not, is the navmesh generator sourec accessible? If at least the navmesh file was edittable... What format is the data stored into it?
